src/ds/ptrwrap.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+namespace snmalloc
+{
+  /**
+   * A convenience type that we can use to annotate internal functions
+   * which return pointers that are headed out to the application or
+   * which have come back.
+   */
+  class ReturnPtr
+  {
+  public:
+    void* unsafe_return_ptr;
```

Comment:
If there are no private members, this should be a `struct`.  That said, can't `unsafe_return_ptr` be a `private` member, with the small handful of functions that are actually allowed to use this declared as `friend`s?  At the moment, if someone accesses the raw pointer value we don't get any kind of warning.

Reply:
Same question for the two other smart pointers.